### PR TITLE
Add test for api/v2/organizations

### PR DESCRIPTION
**Pull request by kbrock**
fixes #9868
Add tests to the v2 api's `organizations_controller_test`

### CODE REVIEW

**Comment by dLobatog** [comment]
[test]

File: test/functional/api/v2/organization_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    get :index, { }
+    assert_response :success
+    assert_not_nil assigns(:organizations)
+  end
+
+  test "should show organization" do
+    get :show, { :id => taxonomies(:organization1).to_param }
+    assert_response :success
+    show_response = ActiveSupport::JSON.decode(@response.body)
+    assert !show_response.empty?
```
Comment (dLobatog) [inline test/functional/api/v2/organization_controller_test.rb:20]: We are changing the standard to refute_empty or refute instead of `assert` negative condition, would you mind to change it?

File: test/functional/api/v2/organization_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  test "should return an error for no params" do
+    post :create
+    assert_response :unprocessable_entity
+  end
+
+  test "should create valid organization" do
+    post :create, { :organization => { :name => "Test Organization" } }
+    assert_response :success
+    show_response = ActiveSupport::JSON.decode(@response.body)
+    assert !show_response.empty?
```
Comment (dLobatog) [inline test/functional/api/v2/organization_controller_test.rb:43]: ditto

File: test/functional/api/v2/organization_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    #assert child nodes are included in response'
+    NODES = ["users", "smart_proxies", "subnets", "compute_resources", "media", "config_templates",
+             "domains", "environments", "hostgroups", "locations", "parameters"].sort
+    NODES.each do |node|
+      assert show_response.keys.include?(node), "'#{node}' child node should be in response but was not"
+    end
+  end
+
+  test "should not create invalid organization" do
+    post :create, { :organization => { :name => "" } }
```
Comment (dLobatog) [inline test/functional/api/v2/organization_controller_test.rb:30]: Covered by https://github.com/theforeman/foreman/blob/develop/test/unit/organization_test.rb#L8

File: test/functional/api/v2/organization_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  end
+
+  test "should update organization on if valid is organization" do
+    ignore_types = ["Domain", "Hostgroup", "Environment", "User", "Medium", "Subnet", "SmartProxy", "ConfigTemplate", "ComputeResource", "Realm"]
+    put :update, { :id => @organization.to_param, :organization => { :name => "New Organization", :ignore_types => ignore_types } }
+    assert_equal "New Organization", Organization.find(@organization.id).name
+    assert_response :success
+  end
+
+  test "should not update invalid organization" do
```
Comment (dLobatog) [inline test/functional/api/v2/organization_controller_test.rb:53]: Covered in https://github.com/theforeman/foreman/blob/develop/test/unit/organization_test.rb#L8

File: test/functional/api/v2/organization_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+  end
+
+  test "should destroy organization if hosts do not use it" do
+    assert_difference('Organization.count', -1) do
+      delete :destroy, { :id => taxonomies(:organization2).to_param }
+    end
+    assert_response :success
+  end
+
+  test "should NOT destroy organization if hosts use it" do
```
Comment (dLobatog) [inline test/functional/api/v2/organization_controller_test.rb:65]: I'm unsure, but maybe this and the `should destroy` test should just instead be unit tests that ensure we check for hosts? Thoughts?

File: test/functional/api/v2/organization_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    FactoryGirl.create(:host, :organization => taxonomies(:organization1))
+    assert_difference('Organization.count', 0) do
+      delete :destroy, { :id => taxonomies(:organization1).to_param }
+    end
+    assert_response :unprocessable_entity
+  end
+
+  test "should update *_ids. test for domain_ids" do
```
Comment (dLobatog) [inline test/functional/api/v2/organization_controller_test.rb:73]: Maybe change this to `should only update ids for "unignored" attributes"?

File: test/functional/api/v2/organization_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+  # using Organization as class to pagination and search metadata
+  # create 26 organizations per name per letter A-Z
+  def add_organizations
+    Array('a'..'z').each do |letter|
+      Organization.create(:name => letter)
+    end
+  end
+
+  test "should return correct metadata if no params passed" do
```
Comment (dLobatog) [inline test/functional/api/v2/organization_controller_test.rb:165]: Could you wrap this in a context `metadata` so that you make a setup block to `add_organizations`? Is creating 29 organizations absolutely necessary? It's going to hit the db really heavily (read: slow tests), please look into how to mock these results, or work with expectations for metadata.

File: test/functional/api/v2/organization_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,243 @@
+require 'test_helper'
+
+class Api::V2::OrganizationsControllerTest < ActionController::TestCase
+  def setup
+    @organization = taxonomies(:organization1)
+    @organization.location_ids = [taxonomies(:location1).id]
+    Rabl.configuration.use_controller_name_as_json_root = false
+  end
+
+  test "should get index" do
```
Comment (dLobatog) [inline test/functional/api/v2/organization_controller_test.rb:10]: Would you mind creating a context per action?

**Comment by dLobatog** [comment]
Thanks @kbrock , tests are sorely needed. 

Just one thought, I saw these are basically coming from the equivalent in locations_controller. If you could DIY this so that we run these tests for both controllers in one file it'd be great. They are 100% equivalent in Foreman.

File: test/functional/api/v2/organization_controller_test.rb (hunk 1 above)
Comment (kbrock) [inline test/functional/api/v2/organization_controller_test.rb:20]: @eLobato sure thing.
you want me to make the same change on the location side?
(would that go into a separate PR?)

**Comment by kbrock** [comment]
@eLobato Fixing this controller test (and probably the locations as well) to be more streamlined is easy.
But addressing the higher level question of why duplicate the same test across 2 controllers is more tricky. Any suggestions on that? 1) try and test the mixin itself. 2) mixin for the test itself

any pointers to other code that shares across tests would be helpful

(I want to add this test because I thought the metadata around the organizations in the [docs](http://theforeman.org/api/apidoc/v2/organizations.html) is lacking. But I guess it is only missing the examples there.)

File: test/functional/api/v2/organization_controller_test.rb (hunk 1 above)
Comment (dLobatog) [inline test/functional/api/v2/organization_controller_test.rb:20]: It can be the same PR

**Comment by tbrisker** [comment]
@kbrock,
Thank you very much for your efforts! 
This PR has been open for a very long time with no activity. 
Please indicate if you plan to continue working on it or not; if not, feel free to close it. 
If you do not respond either way within a couple of weeks, I will assume this PR is abandoned and will go ahead and close it for you.

**Comment by kbrock** [comment]
Thanks for all the help. Sorry to not follow through with this more
